src/widgets/graph.rs

```diff
@@ -0,0 +1,105 @@
+extern crate num;
+use config::Config;
+use widget::State;
+use serde_json::value::Value;
+use super::super::widget::I3BarWidget;
+use num::{ToPrimitive, clamp};
+
+#[derive(Clone, Debug)]
+pub struct GraphWidget {
+    content: Option<String>,
+    icon: Option<String>,
+    state: State,
+    rendered: Value,
+    cached_output: Option<String>,
+    config: Config,
+}
+
+impl GraphWidget {
+    pub fn new(config: Config) -> Self {
+        GraphWidget {
+            content: None,
+            icon: None,
+            state: State::Idle,
+            rendered: json!({
+                "full_text": "",
+                "separator": false,
+                "separator_block_width": 0,
+                "background": "#000000",
+                "color": "#000000"
+            }),
+            config: config,
+            cached_output: None,
+        }
+    }
+
+    pub fn with_icon(mut self, name: &str) -> Self {
+        self.icon = self.config.icons.get(name).cloned();
+        self.update();
+        self
+    }
+
+    pub fn with_state(mut self, state: State) -> Self {
+        self.state = state;
+        self.update();
+        self
+    }
+
+    pub fn set_values<'a, T>(&mut self, content: &'a [T], min: Option<T>, max: Option<T>)
```

Comment:
I swear there was a reason I had that in there, removed it and it was able to elide it fine though. Made another minor change also. should be good now.